include/irr/asset/ICPUVirtualTexture.h

```diff
@@ -0,0 +1,337 @@
+                auto& region = regions->front();
+                region.imageSubresource.mipLevel = 0u;
+                region.imageSubresource.baseArrayLayer = 0u;
+                region.imageSubresource.layerCount = _layers;
+                region.bufferOffset = 0u;
+                region.bufferRowLength = _extent;
+                region.bufferImageHeight = 0u; //tightly packed
+                region.imageOffset = {0u,0u,0u};
+                region.imageExtent = params.extent;
+                auto buffer = core::make_smart_refctd_ptr<ICPUBuffer>(getTexelOrBlockBytesize(_format) * params.extent.width*params.extent.height*params.arrayLayers);
+                image->setBufferAndRegions(std::move(buffer), regions);
+            }
+        }
+
+    private:
+        core::smart_refctd_ptr<ICPUImageView> createView_internal(ICPUImageView::SCreationParams&& _params) const override
+        {
+            return ICPUImageView::create(std::move(_params));
+        }
+    };
+
+    ICPUVirtualTexture(
+        const base_t::IVTResidentStorage::SCreationParams* _residentStorageParams,
+        uint32_t _residentStorageCount,
+        uint32_t _pgTabSzxy_log2 = 8u,
+        uint32_t _pgTabLayers = 256u,
+        uint32_t _pgSzxy_log2 = 7u,
+        uint32_t _tilePadding = 9u,
+        uint32_t _maxAllocatableTexSz_log2 = 14u
+    ) : IVirtualTexture(
+        _pgTabSzxy_log2, _pgTabLayers, _pgSzxy_log2, _tilePadding
+    ) {
+        m_pageTable = createPageTable(_pgTabSzxy_log2, _pgTabLayers, _pgSzxy_log2, _maxAllocatableTexSz_log2);
+        initResidentStorage(_residentStorageParams, _residentStorageCount);
+    }
+
+    STextureData pack(const ICPUImage* _img, const IImage::SSubresourceRange& _subres, ISampler::E_TEXTURE_CLAMP _wrapu, ISampler::E_TEXTURE_CLAMP _wrapv, ISampler::E_TEXTURE_BORDER_COLOR _borderColor) override
+    {
+        const E_FORMAT format = _img->getCreationParameters().format;
+        uint32_t smplrIndex = 0u;
+        ICPUVTResidentStorage* storage = nullptr;
+        {
+            auto found = m_storage.find(getFormatClass(format));
+            if (found==m_storage.end())
+                return STextureData::invalid();
+            storage = static_cast<ICPUVTResidentStorage*>(found->second.get());
+
+            SamplerArray* views = nullptr;
+            if (isFloatingPointFormat(format)||isNormalizedFormat(format)||isScaledFormat(format))
+                views = &m_fsamplers;
+            else if (isSignedFormat(format))
+                views = &m_isamplers;
+            else
+                views = &m_usamplers;
+            auto view_it = std::find_if(views->views->begin(), views->views->end(), [format](const core::smart_refctd_ptr<ICPUImageView>& _view) {return _view->getCreationParameters().format==format;});
+            if (view_it==views->views->end()) //no physical page texture view/sampler for requested format
+                return STextureData::invalid();
```

Comment:
because `EF_RGBA_SRGB` and `EF_RGBA8_UINT` are compatible with each other, I can pack a SRGB image while intending to use a UINT view.

You cannot use the image's format to handle this, `pack` needs to take a list of `viewFormats` that we're going to use the image data with (so we only allocate/make resident once, but we'll have multiple VT entries)